docs/introduction/entities.md

```diff
@@ -356,15 +357,62 @@ class BrandData
+
+If you care about array keys, use a name of the key in the property name in form `propertyByKey`. Eg. [`TransportData::$pricesByCurrencyId`](/packages/framework/src/Model/Transport/TransportData.php).
+
+#### Unknown types
+
+We don't recommend to use `mixed` or `array` phpdoc as they aren't expressive.
+
+The only exception in the framework is `$pluginData` in `CategoryData` and `ProductData`. The phpdoc is an `array` in this case because plugins can contain any type.
+
+#### Entities
+
+It is common that you need to transfer an entity to form or other parts of the system.
+
+If you need to transfer one entity, use phpdoc `entity|null`, eg. `\Shopsys\FrameworkBundle\Model\Pricing\Vat\Vat|null` in [`TransportData::$vat`](/packages/framework/src/Model/Transport/TransportData.php).
+If you need to transfer a collection of entities, use phpdoc `entity[]` and initialize the array in the constructor, eg. `\Shopsys\FrameworkBundle\Model\Payment\Payment[]` in [`TransportData::$payments`](/packages/framework/src/Model/Transport/TransportData.php).
+
+#### Images
+
+To transfer images via the system, use phpdoc `\Shopsys\FrameworkBundle\Component\FileUpload\ImageUploadData` and initialize the field in the constructor as you can see in `BrandData` example above.
+
+#### URL addresses
+
+To transfer URL addresses via the system, use phpdoc `\Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\UrlListData` and initialize the field in the constructor as you can see in `BrandData` example above.
+
+#### Multidomain
+
+[Multidomain property](/docs/introduction/domain-multidomain-multilanguage.md#multidomain-attribute) is an array and has to be indexed by `domainId` - an integer ID of the given domain.
+An example of such property is a `seoH1s` in the `BrandData` example above.
+Data factory has to prepare this array to contain keys for all domain IDs because domain entities are created by these array items (even if their value is null).
+
+Therefore the multidomain field has phpdoc `string[]|null[]` or `int[]|null[]`.
+For boolean multidomain properties, we recommend using default value filled in the factory and phpdoc `bool[]` only, eg. property [`TransportData::$enabled`](/packages/framework/src/Model/Transport/TransportData.php).
+
+#### Multilanguage
 
-As you can see in example above, multidomain or multilanguage field is an array.
+[Multilanguage property](/docs/introduction/domain-multidomain-multilanguage.md#multilanguage-attribute) is an array and has to be indexed by `locale` - a string identifier of language (you can find them in [`domains.yml`](/project-base/app/config/domains.yml)).
+An example of such property is a `descriptions` in the `BrandData` example above.
+Data factory has to prepare this array to contain keys for all locales because translation entities are created by these array items (even if their value is null).
```

Comment:
```suggestion
Data factory has to create items in this array for each locale, otherwise translation entities would not be created correctly (the translation entities should exist for each locale, even with null values).
```